multimodal/src/autogluon/multimodal/learners/default_learner.py

```diff
@@ -0,0 +1,1724 @@
+        self._verify_inference_ready()
+
+        # 2. route to matcher if necessary. TODO: refactor this into matching learner.
+        # We'll leave it as is for now in the predictor.
+        # if self._matcher:
+        #     return self._matcher.evaluate(
+        #         data=data,
+        #         query_data=query_data,
+        #         response_data=response_data,
+        #         id_mappings=id_mappings,
+        #         chunk_size=chunk_size,
+        #         similarity_type=similarity_type,
+        #         cutoffs=cutoffs,
+        #         label=label,
+        #         metrics=metrics,
+        #         return_pred=return_pred,
+        #         realtime=realtime,
+        #     )
+
+        # 3. for detection problems. TODO: refactor this into Detection Learner
+        # i.e. override self._evaluate method in Detection Learner
+        # if self._problem_type == OBJECT_DETECTION:
+        #     if realtime:
+        #         return NotImplementedError(
+        #             f"Current problem type {self._problem_type} does not support realtime predict."
+        #         )
+        #     if isinstance(data, str):
+        #         return evaluate_coco(
+        #             predictor=self,
+        #             anno_file_or_df=data,
+        #             metrics=metrics,
+        #             return_pred=return_pred,
+        #             eval_tool=eval_tool,
+        #         )
+        #     else:
+        #         data = object_detection_data_to_df(data)
+        #         return evaluate_coco(
+        #             predictor=self,
+        #             anno_file_or_df=data,
+        #             metrics=metrics,
+        #             return_pred=return_pred,
+        #             eval_tool="torchmetrics",
+        #         )
+
+        # 4. get return type. TODO: refactor this into sub-learners
+        # i.e. define self.get_return_type() -> str
+        if self._problem_type == NER:
+            ret_type = NER_RET
+        else:
+            ret_type = LOGITS
+
+        # 5. run predict and get outputs
+        outputs = predict(
+            predictor=self,
+            data=data,
+            requires_label=True,
+            realtime=realtime,
+        )
+        logits = extract_from_output(ret_type=ret_type, outputs=outputs)
+
+        # 6. get y_pred and y_true for computing metrics. TODO: refactor this into sub-learners
+        # i.e. define self.transform_logits_to_metric_input(logits)
+        metric_data = {}
+        if self._problem_type in [BINARY, MULTICLASS]:
+            y_pred_prob = logits_to_prob(logits)
+            metric_data[Y_PRED_PROB] = y_pred_prob
+
+        y_pred = self._df_preprocessor.transform_prediction(
+            y_pred=logits,
+            inverse_categorical=False,
+        )
+        y_pred_inv = self._df_preprocessor.transform_prediction(
+            y_pred=logits,
+            inverse_categorical=True,
+        )
```

Comment:
All of this should be moved to an entirely separate method that can take `outputs` + `y` as input. The process of predicting on the data and evaluating should not be hard-coupled.